packages/fuels-core/src/types/transaction_builders.rs

```diff
@@ -252,48 +245,139 @@ impl ScriptTransactionBuilder {
-        let gas_price = self.gas_price.unwrap_or(self.network_info.min_gas_price);
-        let gas_limit = self.gas_limit.unwrap_or(self.network_info.max_gas_per_tx);
+    async fn build_with_provider(self, provider: &impl DryRunner) -> Result<ScriptTransaction> {
+        let is_using_predicates = self.is_using_predicates();
+        let base_offset = if is_using_predicates {
+            self.base_offset()
+        } else {
+            0
+        };
+
+        let num_witnesses = self.num_witnesses()?;
+        let tx = self
+            .resolve_fuel_tx_provider(base_offset, num_witnesses, provider)
+            .await?;
+
+        Ok(ScriptTransaction {
+            tx,
+            is_using_predicates,
+        })
+    }
+
+    fn create_dry_run_witnesses(&self, num_witnesses: u8) -> Vec<Witness> {
+        let unresolved_witnesses_len = self.unresolved_signatures.addr_idx_offset_map.len();
+        repeat(Default::default())
+            // Add one in case there is no witnesses at all
+            .take(num_witnesses as usize + unresolved_witnesses_len + 1)
+            .collect()
+    }
+
+    async fn set_script_gas_limit_to_gas_used(
+        tx: &mut Script,
+        provider: &impl DryRunner,
+        network_info: &NetworkInfo,
+        tolerance: f32,
+    ) -> Result<()> {
+        let consensus_params = &network_info.consensus_parameters;
+        // Add `1` because of rounding
+        let max_gas = tx.max_gas(consensus_params.gas_costs(), consensus_params.fee_params()) + 1;
+
+        // If `gas_limit` was not set use `max_gas_per_tx` but subtract the tx's base maximum cost
+        // TODO: @xgreenx why do I need to / 2
+        tx.set_script_gas_limit((network_info.max_gas_per_tx / 2) - max_gas);
+
+        // Add temp coin for dry run
+        tx.inputs_mut().push(FuelInput::coin_signed(
+            Default::default(),
+            Default::default(),
+            1_000_000_000,
+            Default::default(),
+            TxPointer::default(),
+            0,
+            0u32.into(),
+        ));
+
+        let gas_used = provider
+            .dry_run_and_get_used_gas(tx.clone().into(), tolerance)
+            .await?;
 
+        // Remove temp coin
+        tx.inputs_mut().pop();
+
+        tx.set_script_gas_limit(gas_used);
+
+        Ok(())
+    }
+
+    async fn resolve_fuel_tx_provider(
+        self,
+        base_offset: usize,
+        num_witnesses: u8,
+        provider: &impl DryRunner,
+    ) -> Result<Script> {
+        let mut policies = self.generate_common_fuel_policies();
+        policies.set(
+            PolicyType::WitnessLimit,
+            self.witness_limit.or(Some(DEFAULT_SCRIPT_WITNESS_LIMIT)),
+        );
+
+        let scritp_len = self.script.len();
+        let dry_run_witnesses = self.create_dry_run_witnesses(num_witnesses);
         let mut tx = FuelTransaction::script(
-            gas_price,
-            gas_limit,
-            self.maturity.into(),
+            0, // use temporarily
```

Comment:
If we're temporarily keeping this `0` here, we should create an issue and add a `TODO` here so we don't forget about it.

Reply:
I will improve the comments. 0 is used for the `dry_run`